[BUG] MicrometerObservationCapability not reporting timeouts

Copied from https://github.com/spring-cloud/spring-cloud-openfeign/issues/958
Seems like nobody created an issue here as requested in the original issue 

> **Describe the bug** Feign cloud version: `4.1.0`
> 
> When using the MicrometerObservationCapability, any` java.net.SocketTimeoutException` is not caught and observations is never closed (`MicrometerObservationCapability#enrich only catches FeignExceptions`). Probably affects other errors from the underlying web layer too. The result is that these requests are never added to the metrics. There is either an ordering error, or `MicrometerObservationCapability` should catch more types of errors
> 
> **Sample** No specific config added. Plain FeignClients
> 
> ```
> @FeignClient(name = "NAME", url = "${url.prop}")
> interface Client {
> // calls here
> }
> 
> @Configuration
> class FeignConfig {
>   @Bean
>   def observation(observationRegistry: ObservationRegistry): MicrometerObservationCapability = new MicrometerObservationCapability(observationRegistry, new DefaultFeignObservationConvention())
> }
> ```
> 
> application.yml
> 
> ```
> spring:
>   cloud:
>     openfeign:
>       client:
>         config:
>           default:
>             connect-timeout: 3000
>             read-timeout: 3000
> ```
> 
> Stacktrace
> 
> ```
> Caused by: java.net.SocketTimeoutException: Read timed out
> 	at java.base/sun.nio.ch.NioSocketImpl.timedRead(NioSocketImpl.java:288)
> 	at java.base/sun.nio.ch.NioSocketImpl.implRead(NioSocketImpl.java:314)
> 	at java.base/sun.nio.ch.NioSocketImpl.read(NioSocketImpl.java:355)
> 	at java.base/sun.nio.ch.NioSocketImpl$1.read(NioSocketImpl.java:808)
> 	at java.base/java.net.Socket$SocketInputStream.read(Socket.java:966)
> 	at java.base/java.io.BufferedInputStream.fill(BufferedInputStream.java:244)
> 	at java.base/java.io.BufferedInputStream.read1(BufferedInputStream.java:284)
> 	at java.base/java.io.BufferedInputStream.read(BufferedInputStream.java:343)
> 	at java.base/sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:826)
> 	at java.base/sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:761)
> 	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1688)
> 	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1589)
> 	at java.base/java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:529)
> 	at feign.Client$Default.convertResponse(Client.java:111)
> 	at feign.Client$Default.execute(Client.java:107)
> 	at feign.micrometer.MicrometerObservationCapability.lambda$enrich$1(MicrometerObservationCapability.java:53)
> 	at feign.SynchronousMethodHandler.executeAndDecode(SynchronousMethodHandler.java:100)
> ```

--- a/micrometer/src/main/java/feign/micrometer/MicrometerObservationCapability.java
+++ b/micrometer/src/main/java/feign/micrometer/MicrometerObservationCapability.java
@@ -18,10 +18,10 @@ package feign.micrometer;
 import feign.AsyncClient;
 import feign.Capability;
 import feign.Client;
-import feign.FeignException;
 import feign.Response;
 import io.micrometer.observation.Observation;
 import io.micrometer.observation.ObservationRegistry;
+import java.util.concurrent.CompletionException;
 
 /** Wrap feign {@link Client} with metrics. */
 public class MicrometerObservationCapability implements Capability {
@@ -55,13 +55,15 @@ public class MicrometerObservationCapability implements Capability {
                   this.observationRegistry)
               .start();
 
-      try {
+      try (Observation.Scope scope = observation.openScope()) {
         Response response = client.execute(request, options);
-        finalizeObservation(feignContext, observation, null, response);
+        feignContext.setResponse(response);
         return response;
-      } catch (FeignException ex) {
-        finalizeObservation(feignContext, observation, ex, null);
+      } catch (Throwable ex) {
+        observation.error(ex);
         throw ex;
+      } finally {
+        observation.stop();
       }
     };
   }
@@ -80,24 +82,29 @@ public class MicrometerObservationCapability implements Capability {
                   this.observationRegistry)
               .start();
 
-      try {
+      try (Observation.Scope scope = observation.openScope()) {
         return client
             .execute(feignContext.getCarrier(), options, context)
-            .whenComplete((r, ex) -> finalizeObservation(feignContext, observation, ex, r));
-      } catch (FeignException ex) {
-        finalizeObservation(feignContext, observation, ex, null);
-
+            .whenComplete(
+                (response, ex) -> {
+                  feignContext.setResponse(response);
+                  if (ex != null) {
+                    observation.error(unwrap(ex));
+                  }
+                  observation.stop();
+                });
+      } catch (Throwable ex) {
+        observation.error(ex);
+        observation.stop();
         throw ex;
       }
     };
   }
 
-  private void finalizeObservation(
-      FeignContext feignContext, Observation observation, Throwable ex, Response response) {
-    feignContext.setResponse(response);
-    if (ex != null) {
-      observation.error(ex);
+  private static Throwable unwrap(Throwable ex) {
+    if (ex instanceof CompletionException && ex.getCause() != null) {
+      return ex.getCause();
     }
-    observation.stop();
+    return ex;
   }
 }

--- a/micrometer/src/test/java/feign/micrometer/MicrometerObservationCapabilityTest.java
+++ b/micrometer/src/test/java/feign/micrometer/MicrometerObservationCapabilityTest.java
@@ -1,0 +1,302 @@
+/*
+ * Copyright © 2012 The Feign Authors (feign@commonhaus.dev)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package feign.micrometer;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import feign.AsyncClient;
+import feign.AsyncFeign;
+import feign.Client;
+import feign.Feign;
+import feign.RequestLine;
+import feign.Retryer;
+import feign.Target.HardCodedTarget;
+import io.micrometer.observation.Observation;
+import io.micrometer.observation.ObservationHandler;
+import io.micrometer.observation.tck.TestObservationRegistry;
+import io.micrometer.observation.tck.TestObservationRegistryAssert;
+import java.io.IOException;
+import java.net.SocketTimeoutException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.atomic.AtomicReference;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Behavioural tests for {@link MicrometerObservationCapability} focusing on the cases that the
+ * existing capability did not cover:
+ *
+ * <ul>
+ *   <li>The observation is current (via {@code Observation.Scope}) while the client executes, so
+ *       downstream handlers — most notably tracing handlers that propagate trace/span ids into the
+ *       MDC — see the Feign-scoped observation rather than the parent.
+ *   <li>Exceptions thrown by the underlying client that are <em>not</em> {@code FeignException} —
+ *       {@code IOException}, {@code SocketTimeoutException}, runtime exceptions thrown by a custom
+ *       {@code ErrorDecoder}, and so on — still close the observation and record the error.
+ * </ul>
+ */
+class MicrometerObservationCapabilityTest {
+
+  interface TestClient {
+    @RequestLine("GET /")
+    String get();
+  }
+
+  interface AsyncTestClient {
+    @RequestLine("GET /")
+    CompletableFuture<String> get();
+  }
+
+  private TestObservationRegistry observationRegistry;
+
+  @BeforeEach
+  void setUp() {
+    this.observationRegistry = TestObservationRegistry.create();
+  }
+
+  @Test
+  void scopeIsOpenDuringClientExecution() {
+    AtomicReference<Observation> observedDuringExecute = new AtomicReference<>();
+    Client client =
+        (request, options) -> {
+          observedDuringExecute.set(observationRegistry.getCurrentObservation());
+          return feign.Response.builder()
+              .status(200)
+              .reason("OK")
+              .request(request)
+              .headers(java.util.Collections.emptyMap())
+              .build();
+        };
+
+    TestClient feignClient =
+        Feign.builder()
+            .client(client)
+            .addCapability(new MicrometerObservationCapability(observationRegistry))
+            .target(new HardCodedTarget<>(TestClient.class, "http://localhost"));
+
+    feignClient.get();
+
+    assertThat(observedDuringExecute.get())
+        .as("observation must be active (scoped) while the underlying client executes")
+        .isNotNull();
+    assertThat(observationRegistry.getCurrentObservation())
+        .as("observation must no longer be current after the call returns")
+        .isNull();
+  }
+
+  @Test
+  void recordsNonFeignExceptionThrownByClient() {
+    // Feign's SynchronousMethodHandler wraps IOExceptions, so callers don't see the original
+    // throwable. The observation captured by the capability sits below that wrapping, and is the
+    // only place the underlying error is recorded against the trace.
+    SocketTimeoutException underlying = new SocketTimeoutException("connect timed out");
+
+    Client client =
+        (request, options) -> {
+          throw underlying;
+        };
+
+    TestClient feignClient =
+        Feign.builder()
+            .client(client)
+            .retryer(Retryer.NEVER_RETRY)
+            .addCapability(new MicrometerObservationCapability(observationRegistry))
+            .target(new HardCodedTarget<>(TestClient.class, "http://localhost"));
+
+    assertThatThrownBy(feignClient::get).isInstanceOf(Exception.class);
+
+    TestObservationRegistryAssert.assertThat(observationRegistry)
+        .hasSingleObservationThat()
+        .hasBeenStopped()
+        .hasError(underlying);
+  }
+
+  @Test
+  void recordsRuntimeExceptionThrownByClient() {
+    RuntimeException underlying = new RuntimeException("boom");
+
+    Client client =
+        (request, options) -> {
+          throw underlying;
+        };
+
+    TestClient feignClient =
+        Feign.builder()
+            .client(client)
+            .retryer(Retryer.NEVER_RETRY)
+            .addCapability(new MicrometerObservationCapability(observationRegistry))
+            .target(new HardCodedTarget<>(TestClient.class, "http://localhost"));
+
+    assertThatThrownBy(feignClient::get).isInstanceOf(Exception.class);
+
+    TestObservationRegistryAssert.assertThat(observationRegistry)
+        .hasSingleObservationThat()
+        .hasBeenStopped()
+        .hasError(underlying);
+  }
+
+  @Test
+  void asyncScopeIsOpenDuringClientExecution() {
+    AtomicReference<Observation> observedDuringExecute = new AtomicReference<>();
+    AsyncClient<Object> client =
+        (request, options, context) -> {
+          observedDuringExecute.set(observationRegistry.getCurrentObservation());
+          return CompletableFuture.completedFuture(
+              feign.Response.builder()
+                  .status(200)
+                  .reason("OK")
+                  .request(request)
+                  .headers(java.util.Collections.emptyMap())
+                  .build());
+        };
+
+    AsyncTestClient feignClient =
+        AsyncFeign.<Object>builder()
+            .client(client)
+            .addCapability(new MicrometerObservationCapability(observationRegistry))
+            .target(new HardCodedTarget<>(AsyncTestClient.class, "http://localhost"));
+
+    feignClient.get().join();
+
+    assertThat(observedDuringExecute.get())
+        .as("observation must be active while the async client kicks off the request")
+        .isNotNull();
+  }
+
+  @Test
+  void asyncRecordsNonFeignExceptionFromFailedFuture() {
+    IOException underlying = new IOException("connection reset");
+
+    AsyncClient<Object> client =
+        (request, options, context) -> {
+          CompletableFuture<feign.Response> failed = new CompletableFuture<>();
+          failed.completeExceptionally(underlying);
+          return failed;
+        };
+
+    AsyncTestClient feignClient =
+        AsyncFeign.<Object>builder()
+            .client(client)
+            .retryer(Retryer.NEVER_RETRY)
+            .addCapability(new MicrometerObservationCapability(observationRegistry))
+            .target(new HardCodedTarget<>(AsyncTestClient.class, "http://localhost"));
+
+    assertThatThrownBy(() -> feignClient.get().join()).isInstanceOf(CompletionException.class);
+
+    TestObservationRegistryAssert.assertThat(observationRegistry)
+        .hasSingleObservationThat()
+        .hasBeenStopped()
+        .hasError(underlying);
+  }
+
+  @Test
+  void asyncRecordsSynchronousExceptionFromClient() {
+    RuntimeException underlying = new RuntimeException("immediate failure");
+
+    AsyncClient<Object> client =
+        (request, options, context) -> {
+          throw underlying;
+        };
+
+    AsyncTestClient feignClient =
+        AsyncFeign.<Object>builder()
+            .client(client)
+            .retryer(Retryer.NEVER_RETRY)
+            .addCapability(new MicrometerObservationCapability(observationRegistry))
+            .target(new HardCodedTarget<>(AsyncTestClient.class, "http://localhost"));
+
+    assertThatThrownBy(feignClient::get).isInstanceOf(RuntimeException.class);
+
+    TestObservationRegistryAssert.assertThat(observationRegistry)
+        .hasSingleObservationThat()
+        .hasBeenStopped()
+        .hasError(underlying);
+  }
+
+  @Test
+  void parentObservationIsRestoredAfterCall() {
+    Observation parent = Observation.start("parent", observationRegistry);
+    try (Observation.Scope ignored = parent.openScope()) {
+
+      Client client =
+          (request, options) ->
+              feign.Response.builder()
+                  .status(200)
+                  .reason("OK")
+                  .request(request)
+                  .headers(java.util.Collections.emptyMap())
+                  .build();
+
+      TestClient feignClient =
+          Feign.builder()
+              .client(client)
+              .addCapability(new MicrometerObservationCapability(observationRegistry))
+              .target(new HardCodedTarget<>(TestClient.class, "http://localhost"));
+
+      feignClient.get();
+
+      assertThat(observationRegistry.getCurrentObservation())
+          .as("parent observation must still be current after the Feign call completes")
+          .isSameAs(parent);
+    } finally {
+      parent.stop();
+    }
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  void scopeReceivesObservationCarryingFeignContext() {
+    AtomicReference<Observation.Context> seenContext = new AtomicReference<>();
+
+    observationRegistry
+        .observationConfig()
+        .observationHandler(
+            new ObservationHandler<Observation.Context>() {
+              @Override
+              public void onScopeOpened(Observation.Context context) {
+                seenContext.set(context);
+              }
+
+              @Override
+              public boolean supportsContext(Observation.Context context) {
+                return true;
+              }
+            });
+
+    Client client =
+        (request, options) ->
+            feign.Response.builder()
+                .status(200)
+                .reason("OK")
+                .request(request)
+                .headers(java.util.Collections.emptyMap())
+                .build();
+
+    TestClient feignClient =
+        Feign.builder()
+            .client(client)
+            .addCapability(new MicrometerObservationCapability(observationRegistry))
+            .target(new HardCodedTarget<>(TestClient.class, "http://localhost"));
+
+    feignClient.get();
+
+    assertThat(seenContext.get())
+        .as("scope must propagate the FeignContext to ObservationHandler#onScopeOpened")
+        .isInstanceOf(FeignContext.class);
+  }
+}

